### DISCUSSION TITLE
Different backend possible?

### DISCUSSION THREAD

**Discussion by bbernhard** [wkentaro/labelme#1137] [Ideas / Feature requests]
Hi,

I've been following this project for a while now and I am really impressed by labelme's feature set and how well it is maintained. Awesome work, Kentaro! It's always a pleasure to stumble across an open source project that's so well maintained as yours! 

I am myself the maintainer of [ImageMonkey](https://imagemonkey.io). ImageMonkey is a public open source image dataset with now over 100k CC0 licensed images, ~100k annotated objects and over 140k labeled objects. 

At ImageMonkey's we have a simple online label/annotation editor, but it's not nearly as powerful and feature-rich as labelme. So, while browsing through labelme's sourcecode, I was wondering whether it would be possible to extend labelme with a different backend (kind of like a plugin system). So, instead of always writing the labels/annotations to a file, it would be possible to use a different persistent storage (e.g REST API calls to a service). The backend type can then be selected in the labelme settings, allowing users to switch between different backends. 

I've already looked a bit through the code in order to get an impression where changes would be needed, but I obviously do not know the code base as well as you. I haven't thought much about the interface yet, but maybe something as simple as this could do the trick?

* backend class (abstract base interface which every backend implementation needs to inherit from)
```
class Backend(object):
  def __init__(self):
    pass
  def load(self, args):
    raise NotImplementedError("please implement")
  def persist(self, args):
    raise NotImplementedError("please implement")

  ... 
```

* file based backend
```
class FileBasedBackend(Backend):
  def __init__(self):
    super().__init__()
  def load(self, args):
    #TODO: implement
  def persist(self, args):
    #TODO: implement
```

* A different backend (e.g ImageMonkey Backend)
```
class AnotherBackend(Backend):
  def __init__(self):
    super().__init__()
  def load(self, args):
    #TODO: implement
  def persist(self, args):
    #TODO: implement
```

What do you think about that? Do you think it would be possible to hide the filesystem access behind an interface or would that be too much work?

I've been playing with the idea of forking labelme and giving it a try, but I am a bit afraid that I'll bite off more than I can chew. What do you think? And would you be okay with merging those type of changes back into your codebase or would that change be too big? 

Would love to hear your opinions on that. 

Thanks,
Bernhard

**Comment by wkentaro** [comment]
I think this is possible, and a good direction to make this annotation tool general to different image and annotation storage services.
Mainly image/annotation loading and dumping is implemented in `labelme.LabelFile` class: https://github.com/wkentaro/labelme/blob/master/labelme/label_file.py
so the class for `ImageMonkey` backend will be a similar interface.
